l10n_it_fatturapa_out/wizard/efattura.py

```diff
@@ -145,12 +145,20 @@ def format_quantity(line):
             res = ("{qta:.{precision}f}".format(qta=quantity, precision=uom_precision),)
             return res[0]
 
-        def get_vat_number(vat):
-            # return vat[2:].replace(' ', '') if vat else ""
-            return vat[2:] if vat else ""
-
-        def get_vat_country(vat):
-            return vat[:2].upper() if vat else ""
+        def get_id_fiscale_iva(partner):
+            id_paese = partner.country_id.code
+
+            if partner.vat:
+                if partner.country_id.code == "IT" and partner.vat.startswith("IT"):
+                    id_codice = partner.vat[2:]
+                else:
+                    id_codice = partner.vat
+            else:
+                id_codice = "99999999999"
```

Comment:
Prima di queste modifiche, il codice **99999999999** veniva applicato quando il partner non era **in_eu** mentre ora viene applicato quando non ha `vat`, come mai questa modifica?

Potrebbe avere a che fare con il tuo buon proposito di aggiungere un test (https://github.com/OCA/l10n-italy/pull/2602#issuecomment-1085566131)?

O forse è solo questa la modifica funzionale necessaria per correggere https://github.com/OCA/l10n-italy/issues/2599 e tutto il resto è refactoring? Nel caso, ti chiederei di separare le due cose.

Non ho trovato nelle specifiche della FE dove viene menzionato questo codice (il link http://www.fatturapa.gov.it/export/fatturazione/it/normativa/f-2.htm nel README non viene trovato), riesci mica a trovare un riferimento? Anche nella issue https://github.com/OCA/l10n-italy/issues/2599 vedo solo link a Discord.

Reply:
Il codice 99999999999 credo sia una mera convenzione, il campo è obbligatorio sempre ma il valore non viene controllato dallo SdI.
Sicuramente copre il caso di cliente italiano privo di IVA.

Tuttavia, il resto non si tratta di refactoring e basta.

La PR risolve anche i casi:
- cliente estero non EU con identificativo fiscale NNNNNNNNNNN (senza country code come prefisso), tipo Stati Uniti se non ricordo male;
- cliente estero non EU, con VAT rilasciata da EU, formato EUNNNNNNNNNNN (EU non è un CC valido);
- cliente estero EU, con VAT in formato europeo, ma in Grecia e Irlanda del Nord, alle quali l'EU attribuisce il prefisso VAT in formato 'ELNNNNNNNNNNN' e 'IXNNNNNNNNNNN' rispettivamente, pur essendo i due CC 'GR' e 'GB', odoo produceva XML IdPaese sbagliato.

Lo fa uniformando e semplificando le varie casistiche, per es. prendendo sempre il contry code dall'anagrafica e mai dalla VAT (dove per VAT corrette il CC potrebbe essere sbagliato).